pkg/providers/instancetype/instancetype.go

```diff
@@ -112,91 +130,63 @@ func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1.EC2NodeClass)
 		return s.Zone, s.ZoneID
 	})
-	result := lo.Map(p.instanceTypesInfo, func(i ec2types.InstanceTypeInfo, _ int) *cloudprovider.InstanceType {
-		InstanceTypeVCPU.Set(float64(lo.FromPtr(i.VCpuInfo.DefaultVCpus)), map[string]string{
-			instanceTypeLabel: string(i.InstanceType),
-		})
-		InstanceTypeMemory.Set(float64(lo.FromPtr(i.MemoryInfo.SizeInMiB)*1024*1024), map[string]string{
-			instanceTypeLabel: string(i.InstanceType),
-		})
-
-		zoneData := lo.Map(allZones.UnsortedList(), func(zoneName string, _ int) ZoneData {
-			if !p.instanceTypesOfferings[string(i.InstanceType)].Has(zoneName) || !subnetZones.Has(zoneName) {
-				return ZoneData{
-					Name:      zoneName,
-					Available: false,
-				}
-			}
-			return ZoneData{
-				Name:      zoneName,
-				ID:        subnetZoneToID[zoneName],
-				Available: true,
-			}
-		})
-
-		it := p.instanceTypesResolver.Resolve(ctx, i, zoneData, nodeClass)
+	return lo.Map(p.instanceTypesInfo, func(info ec2types.InstanceTypeInfo, _ int) *cloudprovider.InstanceType {
+		it := p.instanceTypesResolver.Resolve(ctx, info, p.instanceTypesOfferings[string(info.InstanceType)].UnsortedList(), zonesToZoneIDs, nodeClass)
 		if cached, ok := p.discoveredCapacityCache.Get(fmt.Sprintf("%s-%016x", it.Name, amiHash)); ok {
 			it.Capacity[corev1.ResourceMemory] = cached.(resource.Quantity)
 		}
-		for _, of := range it.Offerings {
-			InstanceTypeOfferingAvailable.Set(float64(lo.Ternary(of.Available, 1, 0)), map[string]string{
-				instanceTypeLabel: it.Name,
-				capacityTypeLabel: of.Requirements.Get(karpv1.CapacityTypeLabelKey).Any(),
-				zoneLabel:         of.Requirements.Get(corev1.LabelTopologyZone).Any(),
-			})
-			InstanceTypeOfferingPriceEstimate.Set(of.Price, map[string]string{
-				instanceTypeLabel: it.Name,
-				capacityTypeLabel: of.Requirements.Get(karpv1.CapacityTypeLabelKey).Any(),
-				zoneLabel:         of.Requirements.Get(corev1.LabelTopologyZone).Any(),
-			})
-		}
+		InstanceTypeVCPU.Set(float64(lo.FromPtr(info.VCpuInfo.DefaultVCpus)), map[string]string{
+			instanceTypeLabel: string(info.InstanceType),
+		})
+		InstanceTypeMemory.Set(float64(lo.FromPtr(info.MemoryInfo.SizeInMiB)*1024*1024), map[string]string{
+			instanceTypeLabel: string(info.InstanceType),
+		})
 		return it
 	})
-	p.instanceTypesCache.SetDefault(key, result)
-	return result, nil
 }
 
 func (p *DefaultProvider) UpdateInstanceTypes(ctx context.Context) error {
 	// DO NOT REMOVE THIS LOCK ----------------------------------------------------------------------------
 	// We lock here so that multiple callers to getInstanceTypeOfferings do not result in cache misses and multiple
 	// calls to EC2 when we could have just made one call.
-	// TODO @joinnis: This can be made more efficient by holding a Read lock and only obtaining the Write if not in cache
 	p.muInstanceTypesInfo.Lock()
 	defer p.muInstanceTypesInfo.Unlock()
 
-	var instanceTypes []ec2types.InstanceTypeInfo
-
+	instanceTypes := []ec2types.InstanceTypeInfo{}
```

Comment:
nit: go style prefers using `var instanceTypes []ec2types.InstanceTypeInfo` for empty slice declarations